File: avitowatcher/browser.py
"""Запасной способ получить страницу — через настоящий браузер.

Авито иногда отдаёт защитную заглушку на обычные HTTP-запросы. Тогда страницу
забираем уже установленным в системе Chromium-браузером (Edge есть в Windows 11
из коробки): запускаем его в фоновом режиме с включённым отладочным портом и
управляем по протоколу Chrome DevTools. Так exe остаётся лёгким — Chromium
внутрь не кладём и лишних библиотек не тянем.

Ключ --dump-dom здесь не годится: msedge.exe собран как оконное приложение и
в канал вывода ничего не пишет, поэтому DOM забираем через отладочный порт.
"""
from __future__ import annotations

import base64
import json
import logging
import os
import secrets as _secrets
import socket
import struct
import subprocess
import sys
import time
import urllib.request
from pathlib import Path

from .paths import cache_dir

log = logging.getLogger(__name__)

# Порядок важен: сначала Edge (есть всегда), потом Chrome, потом Brave.
_CANDIDATES = [
    r"%ProgramFiles(x86)%\Microsoft\Edge\Application\msedge.exe",
    r"%ProgramFiles%\Microsoft\Edge\Application\msedge.exe",
    r"%ProgramFiles%\Google\Chrome\Application\chrome.exe",
    r"%ProgramFiles(x86)%\Google\Chrome\Application\chrome.exe",
    r"%LocalAppData%\Google\Chrome\Application\chrome.exe",
    r"%ProgramFiles%\BraveSoftware\Brave-Browser\Application\brave.exe",
]

_CREATE_NO_WINDOW = 0x08000000  # чтобы не мигало консольное окно

USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36"
)


def find_browser() -> Path | None:
    """Путь к первому найденному браузеру или None."""
    for pattern in _CANDIDATES:
        path = Path(os.path.expandvars(pattern))
        if "%" not in str(path) and path.exists():
            return path
    return None


def browser_name() -> str:
    path = find_browser()
    return path.name if path else ""


def is_available() -> bool:
    return find_browser() is not None


# --- минимальный клиент WebSocket ----------------------------------------
class _WebSocket:
    """Ровно столько протокола RFC 6455, сколько нужно для DevTools."""

    def __init__(self, url: str, timeout: float = 30.0) -> None:
        rest = url.split("://", 1)[1]
        hostport, _, path = rest.partition("/")
        host, _, port = hostport.partition(":")
        self._sock = socket.create_connection((host, int(port or 80)), timeout=timeout)
        self._sock.settimeout(timeout)
        self._buffer = b""

        key = base64.b64encode(_secrets.token_bytes(16)).decode("ascii")
        handshake = (
            f"GET /{path} HTTP/1.1\r\n"
            f"Host: {hostport}\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            f"Sec-WebSocket-Key: {key}\r\n"
            "Sec-WebSocket-Version: 13\r\n\r\n"
        )
        self._sock.sendall(handshake.encode("ascii"))

        header = b""
        while b"\r\n\r\n" not in header:
            chunk = self._sock.recv(4096)
            if not chunk:
                raise RuntimeError("Браузер закрыл отладочное соединение")
            header += chunk
        head, _, tail = header.partition(b"\r\n\r\n")
        if b"101" not in head.split(b"\r\n")[0]:
            raise RuntimeError("Браузер не принял отладочное соединение")
        self._buffer = tail

    def _read(self, count: int) -> bytes:
        while len(self._buffer) < count:
            chunk = self._sock.recv(65536)
            if not chunk:
                raise RuntimeError("Браузер закрыл отладочное соединение")
            self._buffer += chunk
        data, self._buffer = self._buffer[:count], self._buffer[count:]
        return data

    def send(self, text: str) -> None:
        payload = text.encode("utf-8")
        header = bytearray([0x81])  # FIN + текстовый кадр
        length = len(payload)
        if length < 126:
            header.append(0x80 | length)
        elif length < 65536:
            header.append(0x80 | 126)
            header += struct.pack(">H", length)
        else:
            header.append(0x80 | 127)
            header += struct.pack(">Q", length)
        mask = _secrets.token_bytes(4)
        header += mask
        masked = bytes(b ^ mask[i % 4] for i, b in enumerate(payload))
        self._sock.sendall(bytes(header) + masked)

    def recv(self) -> str:
        """Возвращает следующее текстовое сообщение, склеивая части кадра."""
        message = b""
        while True:
            first, second = self._read(2)
            fin = bool(first & 0x80)
            opcode = first & 0x0F
            length = second & 0x7F
            if length == 126:
                length = struct.unpack(">H", self._read(2))[0]
            elif length == 127:
                length = struct.unpack(">Q", self._read(8))[0]
            if second & 0x80:                      # сервер маскировать не должен
                mask = self._read(4)
                raw = self._read(length)
                payload = bytes(b ^ mask[i % 4] for i, b in enumerate(raw))
            else:
                payload = self._read(length)

            if opcode == 0x8:
                raise RuntimeError("Браузер разорвал отладочное соединение")
            if opcode == 0x9:                      # ping — отвечаем pong
                self._sock.sendall(bytes([0x8A, 0x80]) + _secrets.token_bytes(4))
                continue
            if opcode == 0xA:
                continue
            message += payload
            if fin:
                return message.decode("utf-8", errors="replace")

    def close(self) -> None:
        try:
            self._sock.close()
        except OSError:
            pass


# --- управление браузером -------------------------------------------------
class _Session:
    """Запущенный браузер с открытым отладочным портом."""

    def __init__(self, exe: Path, profile: Path, proxy: str, timeout: float) -> None:
        self._deadline = time.monotonic() + timeout
        args = [
            str(exe),
            "--headless=new",
            "--disable-gpu",
            "--disable-extensions",
            "--disable-background-networking",
            "--disable-sync",
            "--disable-blink-features=AutomationControlled",
            "--no-first-run",
            "--no-default-browser-check",
            "--mute-audio",
            "--window-size=1440,900",
            "--lang=ru-RU",
            f"--user-agent={USER_AGENT}",
            f"--user-data-dir={profile}",
            "--remote-debugging-port=0",
        ]
        if proxy:
            args.append(f"--proxy-server={proxy}")
        args.append("about:blank")

        self._process = subprocess.Popen(
            args,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            creationflags=_CREATE_NO_WINDOW if sys.platform == "win32" else 0,
        )
        self._port = self._wait_for_port(profile)

    def _wait_for_port(self, profile: Path) -> int:
        """Браузер записывает выбранный порт в файл DevToolsActivePort.

        Завершение запущенного процесса тут ни о чём не говорит: msedge.exe
        перезапускает сам себя и первый процесс сразу выходит с кодом 0,
        пока настоящий браузер продолжает подниматься. Поэтому ждём именно
        файл с портом, а не жизнь процесса.
        """
        marker = profile / "DevToolsActivePort"
        while time.monotonic() < self._deadline:
            try:
                first_line = marker.read_text("utf-8").splitlines()[0]
                return int(first_line)
            except (OSError, IndexError, ValueError):
                time.sleep(0.2)
        raise RuntimeError("Браузер не открыл отладочный порт за отведённое время")

    def _http(self, path: str) -> object:
        url = f"http://127.0.0.1:{self._port}{path}"
        with urllib.request.urlopen(url, timeout=15) as response:
            return json.loads(response.read().decode("utf-8"))

    def page_socket_url(self) -> str:
        while time.monotonic() < self._deadline:
            try:
                targets = self._http("/json/list")
            except Exception:
                time.sleep(0.2)
                continue
            for target in targets:
                if target.get("type") == "page" and target.get("webSocketDebuggerUrl"):
                    return target["webSocketDebuggerUrl"]
            try:
                created = self._http("/json/new?about:blank")
                if created.get("webSocketDebuggerUrl"):
                    return created["webSocketDebuggerUrl"]
            except Exception:
                pass
            time.sleep(0.3)
        raise RuntimeError("Не удалось открыть вкладку в браузере")

    @property
    def deadline(self) -> float:
        return self._deadline

    def close(self) -> None:
        """Закрывает браузер.

        Убить запущенный процесс мало — он давно вышел, а работает его
        потомок. Поэтому просим сам браузер закрыться по отладочному
        протоколу; нашей копии это касается, чужие окна не трогаем.
        """
        try:
            version = self._http("/json/version")
            socket_url = version.get("webSocketDebuggerUrl")
            if socket_url:
                websocket = _WebSocket(socket_url, timeout=10)
                try:
                    websocket.send(json.dumps({"id": 1, "method": "Browser.close"}))
                    time.sleep(0.3)
                finally:
                    websocket.close()
        except Exception as exc:
            log.debug("Не удалось закрыть браузер штатно: %s", exc)
        if self._process.poll() is None:
            self._process.terminate()
            try:
                self._process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._process.kill()


def fetch(url: str, timeout: int = 60, proxy: str = "") -> str:
    """Загружает страницу браузером и возвращает HTML после исполнения скриптов."""
    exe = find_browser()
    if exe is None:
        raise RuntimeError(
            "Не найден Chromium-браузер (Edge/Chrome). "
            "Установите Microsoft Edge или Google Chrome."
        )

    profile = cache_dir() / "browser-profile"
    profile.mkdir(parents=True, exist_ok=True)
    # Файл с портом остаётся от прошлого запуска — иначе подключимся к старому.
    (profile / "DevToolsActivePort").unlink(missing_ok=True)

    session = _Session(exe, profile, proxy, float(timeout))
    websocket: _WebSocket | None = None
    try:
        websocket = _WebSocket(session.page_socket_url(), timeout=float(timeout))
        message_id = 0

        def call(method: str, params: dict | None = None) -> dict:
            nonlocal message_id
            message_id += 1
            current = message_id
            websocket.send(json.dumps(
                {"id": current, "method": method, "params": params or {}}
            ))
            while True:
                data = json.loads(websocket.recv())
                if data.get("id") == current:
                    if "error" in data:
                        raise RuntimeError(data["error"].get("message", method))
                    return data.get("result", {})

        def wait_for(event: str, until: float) -> bool:
            while time.monotonic() < until:
                try:
                    data = json.loads(websocket.recv())
                except socket.timeout:
                    return False
                if data.get("method") == event:
                    return True
            return False

        call("Page.enable")
        call("Page.navigate", {"url": url})
        wait_for("Page.loadEventFired", min(session.deadline, time.monotonic() + timeout))
        time.sleep(1.5)  # даём дорисоваться тому, что подгружается скриптами

        # Фотографии нижних карточек грузятся только когда до них долистают,
        # поэтому прокручиваем страницу до конца и возвращаемся наверх.
        try:
            for step in range(1, 5):
                call("Runtime.evaluate", {
                    "expression": f"window.scrollTo(0, document.body.scrollHeight*{step}/4)",
                })
                time.sleep(0.6)
            call("Runtime.evaluate", {"expression": "window.scrollTo(0, 0)"})
            time.sleep(0.4)
        except RuntimeError:
            pass

        result = call("Runtime.evaluate", {
            "expression": "document.documentElement.outerHTML",
            "returnByValue": True,
        })
        html = result.get("result", {}).get("value") or ""
    finally:
        if websocket is not None:
            websocket.close()
        session.close()

    if len(html) < 500:
        raise RuntimeError("Браузер вернул пустую страницу")
    return html


def self_test() -> str:
    """Проверка запасного режима — используется кнопкой в настройках."""
    html = fetch("https://www.avito.ru/", timeout=90)
    return f"{browser_name()}: получено {len(html)} символов"

File: src/net/browser.cpp
#include "net/browser.h"

#include <windows.h>

#include <chrono>
#include <thread>

#include "net/http.h"
#include "net/websocket.h"
#include "util/json.h"
#include "util/log.h"
#include "util/paths.h"
#include "util/strings.h"

namespace net {
namespace {

using util::Json;

// Edge first: it is always present on Windows 11.
const wchar_t* kCandidates[] = {
    L"%ProgramFiles(x86)%\\Microsoft\\Edge\\Application\\msedge.exe",
    L"%ProgramFiles%\\Microsoft\\Edge\\Application\\msedge.exe",
    L"%ProgramFiles%\\Google\\Chrome\\Application\\chrome.exe",
    L"%ProgramFiles(x86)%\\Google\\Chrome\\Application\\chrome.exe",
    L"%LocalAppData%\\Google\\Chrome\\Application\\chrome.exe",
    L"%ProgramFiles%\\BraveSoftware\\Brave-Browser\\Application\\brave.exe",
};

constexpr const char* kUserAgent =
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36";

std::wstring expand(const wchar_t* pattern) {
    wchar_t buffer[MAX_PATH * 2] = {};
    DWORD length = ExpandEnvironmentStringsW(pattern, buffer, MAX_PATH * 2);
    if (length == 0 || length > MAX_PATH * 2) return {};
    return std::wstring(buffer, length - 1);
}

void sleep_ms(int milliseconds) {
    std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
}

double monotonic_seconds() {
    return static_cast<double>(GetTickCount64()) / 1000.0;
}

// The browser relaunches itself and the process we spawned exits right away, so
// the only reliable way to stop our instance is to ask it over the protocol.
class BrowserProcess {
public:
    ~BrowserProcess() { stop(); }

    bool start(const std::wstring& exe, const std::wstring& profile,
               const std::string& proxy, double deadline, std::string& error) {
        profile_ = profile;
        util::ensure_dir(profile);
        // A stale port file would point at a browser that is already gone.
        util::delete_file(profile + L"\\DevToolsActivePort");

        std::wstring command =
            L"\"" + exe + L"\""
            L" --headless=new"
            L" --disable-gpu"
            L" --disable-extensions"
            L" --disable-background-networking"
            L" --disable-sync"
            L" --disable-blink-features=AutomationControlled"
            L" --no-first-run"
            L" --no-default-browser-check"
            L" --mute-audio"
            L" --window-size=1440,900"
            L" --lang=ru-RU"
            L" --user-agent=\"" + util::widen(kUserAgent) + L"\""
            L" --user-data-dir=\"" + profile + L"\""
            L" --remote-debugging-port=0";
        if (!proxy.empty()) command += L" --proxy-server=\"" + util::widen(proxy) + L"\"";
        command += L" about:blank";

        STARTUPINFOW startup = {};
        startup.cb = sizeof(startup);
        startup.dwFlags = STARTF_USESHOWWINDOW;
        startup.wShowWindow = SW_HIDE;
        PROCESS_INFORMATION info = {};

        std::vector<wchar_t> mutable_command(command.begin(), command.end());
        mutable_command.push_back(L'\0');

        if (!CreateProcessW(nullptr, mutable_command.data(), nullptr, nullptr, FALSE,
                            CREATE_NO_WINDOW, nullptr, nullptr, &startup, &info)) {
            error = "Не удалось запустить браузер";
            return false;
        }
        CloseHandle(info.hThread);
        process_ = info.hProcess;

        port_ = wait_for_port(deadline);
        if (port_ == 0) {
            error = "Браузер не открыл отладочный порт за отведённое время";
            return false;
        }
        return true;
    }

    int port() const { return port_; }

    void stop() {
        if (port_ > 0) {
            // Ask the browser to close itself; our own user-data-dir means this
            // never touches windows the user opened.
            HttpClient client;
            client.set_timeout_seconds(5);
            Response version = client.get(endpoint("/json/version"));
            if (version.ok) {
                Json root = Json::parse(version.body);
                std::string socket_url = root["webSocketDebuggerUrl"].as_string();
                if (!socket_url.empty()) {
                    WebSocket socket;
                    if (socket.connect(socket_url, 5)) {
                        Json command = Json::object();
                        command["id"] = Json(1);
                        command["method"] = Json("Browser.close");
                        socket.send_text(command.dump());
                        sleep_ms(300);
                        socket.close();
                    }
                }
            }
            port_ = 0;
        }
        if (process_) {
            if (WaitForSingleObject(process_, 0) == WAIT_TIMEOUT) TerminateProcess(process_, 0);
            CloseHandle(process_);
            process_ = nullptr;
        }
    }

    std::string endpoint(const std::string& path) const {
        return "http://127.0.0.1:" + std::to_string(port_) + path;
    }

private:
    // The browser writes the port it picked into DevToolsActivePort. Watching
    // the spawned process is pointless: it exits with code 0 almost at once
    // while the real browser keeps starting up.
    int wait_for_port(double deadline) {
        const std::wstring marker = profile_ + L"\\DevToolsActivePort";
        while (monotonic_seconds() < deadline) {
            std::string text;
            if (util::read_file(marker, text)) {
                size_t newline = text.find('\n');
                std::string first = util::trim(newline == std::string::npos
                                                   ? text
                                                   : text.substr(0, newline));
                int value = std::atoi(first.c_str());
                if (value > 0) return value;
            }
            sleep_ms(200);
        }
        return 0;
    }

    std::wstring profile_;
    HANDLE process_ = nullptr;
    int port_ = 0;
};

}  // namespace

std::wstring find_browser() {
    for (const wchar_t* pattern : kCandidates) {
        std::wstring path = expand(pattern);
        if (!path.empty() && path.find(L'%') == std::wstring::npos &&
            util::file_exists(path)) {
            return path;
        }
    }
    return {};
}

std::string browser_name() {
    std::wstring path = find_browser();
    if (path.empty()) return {};
    size_t slash = path.find_last_of(L"\\/");
    return util::narrow(slash == std::wstring::npos ? path : path.substr(slash + 1));
}

bool browser_available() {
    return !find_browser().empty();
}

BrowserResult browser_fetch(const std::string& url, int timeout_seconds,
                            const std::string& proxy) {
    BrowserResult result;

    std::wstring exe = find_browser();
    if (exe.empty()) {
        result.error =
            "Не найден Chromium-браузер (Edge/Chrome). "
            "Установите Microsoft Edge или Google Chrome.";
        return result;
    }

    const double deadline = monotonic_seconds() + timeout_seconds;
    BrowserProcess browser;
    if (!browser.start(exe, util::browser_profile_dir(), proxy, deadline, result.error)) {
        return result;
    }

    // Find a page target to attach to.
    HttpClient client;
    client.set_timeout_seconds(15);
    std::string socket_url;
    while (monotonic_seconds() < deadline && socket_url.empty()) {
        Response targets = client.get(browser.endpoint("/json/list"));
        if (targets.ok) {
            Json list = Json::parse(targets.body);
            for (size_t i = 0; i < list.size(); ++i) {
                const Json& target = list.at(i);
                if (target["type"].as_string() == "page" &&
                    !target["webSocketDebuggerUrl"].as_string().empty()) {
                    socket_url = target["webSocketDebuggerUrl"].as_string();
                    break;
                }
            }
        }
        if (socket_url.empty()) {
            Response created = client.get(browser.endpoint("/json/new?about:blank"));
            if (created.ok) {
                socket_url = Json::parse(created.body)["webSocketDebuggerUrl"].as_string();
            }
        }
        if (socket_url.empty()) sleep_ms(300);
    }

    if (socket_url.empty()) {
        result.error = "Не удалось открыть вкладку в браузере";
        return result;
    }

    WebSocket socket;
    if (!socket.connect(socket_url, timeout_seconds)) {
        result.error = socket.error();
        return result;
    }

    int next_id = 0;
    auto call = [&](const std::string& method, const Json& params, Json& reply) -> bool {
        int id = ++next_id;
        Json command = Json::object();
        command["id"] = Json(id);
        command["method"] = Json(method);
        command["params"] = params.is_null() ? Json::object() : params;
        if (!socket.send_text(command.dump())) return false;

        while (monotonic_seconds() < deadline) {
            std::string text;
            if (!socket.receive_text(text)) return false;
            Json message = Json::parse(text);
            if (message["id"].as_int(-1) == id) {
                if (message.has("error")) {
                    result.error = message["error"]["message"].as_string("ошибка браузера");
                    return false;
                }
                reply = message["result"];
                return true;
            }
        }
        return false;
    };

    auto wait_for_event = [&](const std::string& event, double until) {
        while (monotonic_seconds() < until) {
            std::string text;
            if (!socket.receive_text(text)) return;
            if (Json::parse(text)["method"].as_string() == event) return;
        }
    };

    Json reply;
    if (!call("Page.enable", Json::object(), reply)) {
        if (result.error.empty()) result.error = "Браузер не принял команду";
        return result;
    }

    Json navigate = Json::object();
    navigate["url"] = Json(url);
    if (!call("Page.navigate", navigate, reply)) {
        if (result.error.empty()) result.error = "Браузер не открыл страницу";
        return result;
    }

    wait_for_event("Page.loadEventFired", std::min(deadline, monotonic_seconds() + timeout_seconds));
    sleep_ms(1500);  // let scripts finish painting

    // Cards below the fold only load their photos once scrolled into view.
    for (int step = 1; step <= 4; ++step) {
        Json scroll = Json::object();
        scroll["expression"] = Json("window.scrollTo(0, document.body.scrollHeight*" +
                                    std::to_string(step) + "/4)");
        Json ignored;
        if (!call("Runtime.evaluate", scroll, ignored)) break;
        sleep_ms(600);
    }
    {
        Json scroll = Json::object();
        scroll["expression"] = Json("window.scrollTo(0, 0)");
        Json ignored;
        call("Runtime.evaluate", scroll, ignored);
        sleep_ms(400);
    }

    Json evaluate = Json::object();
    evaluate["expression"] = Json("document.documentElement.outerHTML");
    evaluate["returnByValue"] = Json(true);
    if (!call("Runtime.evaluate", evaluate, reply)) {
        if (result.error.empty()) result.error = "Не удалось получить содержимое страницы";
        return result;
    }

    result.html = reply["result"]["value"].as_string();
    socket.close();

    if (result.html.size() < 500) {
        result.error = "Браузер вернул пустую страницу";
        return result;
    }
    result.ok = true;
    return result;
}

}  // namespace net
